### PR TITLE
Port all task definitions to task avoidance api

### LINKED ISSUE #56610: Gradle task avoidance api should be used consistently in the build
At several places in the gradle build we still see usage of the old task creation api. This should be replaced by using the task avoidance api of gradle (`tasks.register(...){}`).

This makes the build more consistent and eventually gives us slightly better configuration time. Though this step alone might not result in many task creations being avoided yet as this also requires some further optimisations in how existing tasks are requested within the build scripts and different plugins. This will addressed later on in a different issue.

Update: Many tasks have been converted. Two major remainings:

- RestIntegTest tasks are created eagerly. Tasks of type `RestIntegTest` created during task graph calculation causes 
1. the creation of clusters late after all build scripts have been evaluated.
2. distributions late in the build _after_ the current setupDistribution (see https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java#L82) 

One fix for 1) would be, to configure test clusters that are created by a `RestIntegTest` in the context of the task like: 

```
tasks.named('integTest').configure {
   testclusters.integTest {
       // test cluster configuration
   }
}
```
The benefit of this is that the test cluster configuration would only happen if the cluster is required by the build. 

Fixing 2) is more complex as we need to find a different solution for delaying the distribution setup. We can probably replace the `afterEvaluate`  usage by using `configuration.defaultDependencies{}` hooks

### PR DESCRIPTION
**Pull request by breskeby**
This finishes porting all tasks created in gradle build scripts and plugins to use the task avoidance api (see https://github.com/elastic/elasticsearch/issues/56610) 

Screenshot 2020-12-22 at 13 39 35

As a next step I'll look into enforcing this api in favour of immediate task creation . At the moment this is not provided by gradle and requires a custom solution 

Fixes #56610

### CODE REVIEW

File: test/fixtures/gcs-fixture/build.gradle [hunk 1]
```diff
@@ -26,8 +26,8 @@ dependencies {
   api project(':server')
 }
 
-preProcessFixture {
-  dependsOn jar, configurations.runtimeClasspath
+tasks.named("preProcessFixture").configure {
+  dependsOn tasks.named("jar"), configurations.runtimeClasspath
```
Comment (mark-vieira) [inline test/fixtures/gcs-fixture/build.gradle:30]: Looks like `jar` and `runtimeClasspath` should also be inputs here. In fact we should just do that and then rely on the implicit dependencies.

File: x-pack/plugin/ccr/qa/multi-cluster/build.gradle [hunk 2]
```diff
@@ -9,51 +9,53 @@ dependencies {
...
+    setting 'cluster.remote.leader_cluster.seeds',
+            { "\"${testClusters.named('leader-cluster').get().getAllTransportPortURI().join(",")}\"" }
+  }
 }
 
-testClusters."leader-cluster" {
-  testDistribution = 'DEFAULT'
-  setting 'xpack.license.self_generated.type', 'trial'
+tasks.register("leader-cluster", RestIntegTestTask) {
+  mustRunAfter("precommit")
```
Comment (mark-vieira) [inline x-pack/plugin/ccr/qa/multi-cluster/build.gradle:26]: Does this not require a `useCluster`?

**Review by mark-vieira — APPROVED** [review]
This is so awesome that we've gotten down to **zero** tasks created at configuration time. I wouldn't have thought it possible. Well done.

Just a couple of minor comments, otherwise LGTM.

File: modules/reindex/build.gradle [hunk 3]
```diff
@@ -128,7 +129,7 @@ if (Os.isFamily(Os.FAMILY_WINDOWS)) {
     // see: https://github.com/elastic/elasticsearch/issues/51202
     versions = ['2', '1']
   }
-  for (String version : versions) {
+  versions.each { version ->
```
[the remark this answers is not in the document]
Comment (breskeby) [reply modules/reindex/build.gradle:132]: yes. It fixes exactly the weird issue that you describe here with for loops. I think what we see there is this bug here: https://issues.apache.org/jira/browse/GROOVY-8826

File: x-pack/plugin/ccr/qa/multi-cluster/build.gradle (hunk 2 above)
Comment (breskeby) [reply x-pack/plugin/ccr/qa/multi-cluster/build.gradle:26]: The name matching cluster is always wired here: https://github.com/elastic/elasticsearch/blob/0911d04467ef3f507321a3b92ec831478a5b66cc/buildSrc/src/main/java/org/elasticsearch/gradle/test/RestTestBasePlugin.java#L47

File: test/fixtures/gcs-fixture/build.gradle (hunk 1 above)
Comment (breskeby) [reply test/fixtures/gcs-fixture/build.gradle]: fixed

**Pull request merged**